File: backend/app/schemas/capture.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class CreateSessionRequest(BaseModel):
    source_type: str | None = Field(default=None, examples=["event"])
    source_label: str | None = Field(default=None, examples=["Computex 2026"])


class UpdateSessionRequest(BaseModel):
    source_label: str | None = None
    status: str | None = Field(default=None, examples=["closed"])


class CaptureSessionResponse(BaseModel):
    id: UUID
    user_id: UUID
    workspace_id: UUID
    source_type: str | None
    source_label: str | None
    status: str
    card_count: int
    confirmed_count: int
    pending_count: int
    started_at: datetime
    closed_at: datetime | None

    model_config = {"from_attributes": True}


class DuplicateWarning(BaseModel):
    previous_card_id: UUID
    scanned_at: datetime
    message: str


class RawCardUploadResponse(BaseModel):
    raw_card_id: UUID
    status: str
    capture_session_id: UUID | None
    duplicate_warning: DuplicateWarning | None = None


class OcrResultSummary(BaseModel):
    extracted_fields: dict
    field_confidences: dict
    overall_confidence: float | None = None
    engine: str | None = None
    engine_version: str | None = None

    model_config = {"from_attributes": True}


class CardListItem(BaseModel):
    id: UUID
    status: str
    review_status: str
    capture_method: str
    source_label: str | None
    image_url: str | None
    created_at: datetime
    ocr_summary: OcrResultSummary | None = None

    model_config = {"from_attributes": True}


class CardListResponse(BaseModel):
    items: list[CardListItem]


class CardDetailResponse(BaseModel):
    id: UUID
    capture_session_id: UUID | None
    status: str
    review_status: str
    version: int
    image_url: str | None
    capture_method: str
    source_type: str | None
    source_label: str | None
    created_at: datetime
    ocr_result: OcrResultSummary | None = None

    model_config = {"from_attributes": True}


class ReviewCardRequest(BaseModel):
    name: str | None = None
    company: str | None = None
    title: str | None = None
    version: int


class BatchUploadResponse(BaseModel):
    cards: list[RawCardUploadResponse]

File: frontend/shared/types/capture.ts
export interface CaptureSession {
  id: string;
  user_id: string;
  workspace_id: string;
  source_type: string | null;
  source_label: string | null;
  status: string;
  card_count: number;
  confirmed_count: number;
  pending_count: number;
  started_at: string;
  closed_at: string | null;
}

export interface OcrSummary {
  extracted_fields: Record<string, unknown>;
  field_confidences: Record<string, number>;
  overall_confidence?: number | null;
  engine?: string | null;
  engine_version?: string | null;
}

export interface CardListItem {
  id: string;
  status: string;
  review_status: string;
  review_deferred_at: string | null;
  capture_method: string;
  source_label: string | null;
  image_url: string | null;
  created_at: string;
  version?: number;
  ocr_summary: OcrSummary | null;
}

export interface CardDetail extends CardListItem {
  capture_session_id: string | null;
  version: number;
  source_type: string | null;
  ocr_result: OcrSummary | null;
}

export interface DuplicateWarning {
  previous_card_id: string;
  scanned_at: string;
  message: string;
}

export interface RawCardUploadResponse {
  raw_card_id: string;
  status: string;
  capture_session_id: string | null;
  duplicate_warning?: DuplicateWarning | null;
}

export interface ImportCardResponse {
  raw_card_id: string;
  status: string;
  review_status: string;
  capture_method: string;
  message: string;
  extracted_preview?: {
    name?: string | null;
    company?: string | null;
    title?: string | null;
  } | null;
}

export type ThumbnailStatus = "uploading" | "queued" | "ocr_processing" | "ocr_done" | "ocr_failed" | "upload_failed";

export interface CaptureThumbnail {
  id: string;
  localUrl: string;
  status: ThumbnailStatus;
  cardId?: string;
}
